### ISSUE TITLE
Proposal: Motion-grounded residual world model for FutureState (ego-warp + dynamic residual + visual-history conditioning) Relates to: #13 (FutureState JEPA loss), builds on #80 (loss + action conditioning), #82 (temporal memory).

### ISSUE THREAD

**Issue by intisar1020** [autowarefoundation/auto_e2e#84]
## Summary

This proposes a small, additive refinement to how `FutureState` predicts the future BEV, building on the JEPA feature-reconstruction loss from #13 / #80 and the temporal-memory interface from #82. The idea has two parts:

1. **Motion-grounded residual:**  predict the future BEV as `Warp_ego(z_t) + Δ_dynamic`, where a *parameter-free* geometric warp removes the predictable ego-induced motion and `FutureState` learns only the dynamic residual.
2. **Visual-history conditioning:** condition the dynamic residual on the temporal context from #82, so it can estimate agent velocities that a single frame might nt be able to reveal.

This is a sketch for discussion. I am fairly new to the codebase and may be missing context, so corrections are very welcome would be really helpful. 

## Motivation

With the JEPA loss merged in #80, `FutureState` now has a  training signal (predicted future BEV vs. frozen-backbone targets on the actual future frames). Two observations about the current prediction setup:

1. **`FutureState` conditions on a single current frame** (`fused_features` + `ego_hidden`, + optional action from #80): it has no access to *past* visual context.
2. Between `t` and `t+Δ`, the BEV changes for **two very different reasons**, currently modelled as one learned delta:
   - **ego-induced apparent motion:** the whole BEV shifts/rotates because *we* moved. Large, dominant, and *exactly known* from the ego trajectory. This is really a spatial transform, which an additive delta might model poorly.
   - **dynamic motion:**  other agents actually moving. The genuinely uncertain part.

This connects to the "predict deltas" direction I floated earlier (TDV).

## Part 1  Motion-grounded residual

Decompose the future BEV into a known geometric part and a learned dynamic part:

```
z_{t+Δ} ≈ Warp_ego(z_t)                         + Δ_dynamic
          └─ known ego pose change ─┘   └─ FutureState predicts only this ─┘
```

- **`Warp_ego(z_t)`:**  resample the current BEV into the future ego frame using the *known* ego displacement over each horizon. That displacement is the **future** ego pose change over `[t, t+h]`: at **training** it comes from the **ground-truth future egomotion** (the same future window that builds `trajectory_target`)
- **`Δ_dynamic`:**  `FutureState` predicts only the residual where actors deviate from the static-world warp.

Rationale: a *pure additive* residual (`z_t + Δ`) must absorb both the ego spatial transform and true dynamics with one term. Warping the (exactly known) ego motion out first lets the network focus capacity on dynamics, and gives the JEPA loss a strong prior that **also helps guard against representation collapse** (the warp anchors the prediction). Shapes are unchanged, so the #80 loss, targets, and the training loop (#78 / #14) shouldn't need to change.

Note on inputs: this adds an explicit `ego_motion` argument `(dx, dy, dθ)` per horizon. this is distinct from `ego_hidden`. `ego_motion` is explicit geometry consumed by the warp; `ego_hidden` remains the planner's learned intent vector used as a conditioning bias. They are complementary, not redundant.

## Part 2  Condition the dynamic residual on visual history

From a single frame, **agent velocity is unobservable** . Velocity/acceleration might be only recoverable from multiple past frames, which is exactly what `Δ_dynamic` needs. With the `[B,T,feat]` history interface and the `one_hz` encoder from #82, we can inject a **visual** temporal context into `FutureState`.

Which history? The **visual/scene** context  (not egomotion history), whose information (our own motion) is already handled by `Warp_ego`. Two tiers:

- **Tier 1 (start here):** global `visual_ctx` → per-channel bias added to the BEV before the conv, exactly the pattern `FutureState` already uses for `ego_hidden`. Cheap, drop-in. Conveys scene-level motion context (not per-cell though).
- **Tier 2 (follow-up):** spatial BEV history (past BEV maps via a ConvGRU) for per-cell motion. Richer but memory-heavy. but may be we can compress the feature along channel dimenstion.

## Why this should help

Same argument as #13 and UniAD. A richer, geometrically-grounded self-supervised signal gives the Backbone + Fusion a more direct gradien. Also forcing the model to predict *dynamics* (not ego-motion as its already knows) is a more targeted objective. The introspection idea in #13 still applies

## Open questions

1. **Target alignment**  should the frozen-backbone target also be expressed in the *future* ego frame, so predicted and target share a frame?
2. **Warp source**  use **ground-truth future egomotion** at training (clean, exact) and the **planned trajectory** at inference; should we also use the planned trajectory at training for strict train/test consistency? (The warp needs *future* motion over `[t, t+h]`, so the past `egomotion_history` is only the integration's initial condition, not the source.) Or we can do both 
  - a. Planned-at-train: train=test consistency, but the warp is noisy early in training (when the planner is bad), so the residual target is messier.
  - b. Mixed (best of both): warm up with GT motion, then gradually switch to predicted motion (scheduled sampling).
3. **History conditioning** start with a global per-channel bias (cheap, like `ego_hidden`) and defer spatial per-cell motion (ConvGRU)?
4. **Memory budget**  the forecast is already heavy at 450×300; the warp is cheap but history adds inputs. bf16 / downsampling, per @gcordova10's point (3) in #80.
5. **Loss weighting**  keep equal horizon weights for now (per @m-zain-khawaja in #13), revisit if convergence needs it.

## References

- Daithankar et al., *You Don't Need Strong Assumptions: Visual Representation Learning via Temporal Differences (TDV)*, arXiv:2606.15956
- LeCun, *A Path Towards Autonomous Machine Intelligence (JEPA)*, 2022
- Hu et al., *MILE: Model-Based Imitation Learning for Urban Driving*, NeurIPS 2022
- Grill et al., *Bootstrap Your Own Latent (BYOL)*, NeurIPS 2020
- Hu et al., *Planning-oriented Autonomous Driving (UniAD)*, CVPR 2023
- Hu et al., *GAIA-1: A Generative World Model for Autonomous Driving*, 2023

**Comment by gcordova10** [comment]
Thanks @intisar1020, this is a really nice refinement and I think you're right on the key point: a **pure additive residual `z_t + Δ` has to absorb both the ego spatial transform and the true dynamics with one term**, and the ego part is a *warp*, not an addition. I'd actually merged a plain additive `tdv` mode (`z_{t+k} = z_t + Δz_k`) following your earlier #13 note — your `Warp_ego(z_t) + Δ_dynamic` is the better-motivated version of exactly that, so I'm +1.

To help move it forward I put together a small **prototype** (`predictor_mode="ego_warp"`, stacked on the FutureState changes from the JEPA-training branch):
- a **parameter-free** BEV warp via `affine_grid`+`grid_sample` (differentiable w.r.t. the feature, so the gradient still reaches the backbone), driven by a per-horizon `ego_motion (dx, dy, dθ)`;
- the conv then predicts only `Δ_dynamic_k`, and the output is `Warp_ego(z_t) + Δ_dynamic_k`;
- optional **Tier-1 visual-history conditioning** (a global `visual_context` per-channel bias, mirroring how `ego_hidden` is injected) — cheap drop-in for the agent-velocity signal you describe.

It's additive/opt-in (default mode unchanged) and tested for the convention-independent invariants (zero motion → identity, translation moves content, grad flow). Happy for you to drive this / co-author — it's your idea.

One thing the refactor changes: since the slow World Action Model now reconstructs **multi-view** features (#85), the ego-warp belongs on the **reactive/BEV side** (where the metric BEV grid lives), not in the multi-view slow model. So let's first confirm with @m-zain-khawaja where the BEV future-state lands in `Reactive_E2E` post-refactor, then slot the warp there.

A few things I think we should pin down together (your open questions):
1. **BEV must be metric & ego-centric** for the warp — so this lives on the reactive/BEV path (450×300 @0.4 m, #35), not on the multi-view World Action Model. Agreed?
2. **Axis/sign convention** (which spatial dim is longitudinal vs lateral, sign of yaw) must match the BEV producer (FeatureFusion / #40). My prototype documents one convention but we should confirm it against the actual grid.
3. **Warp source & horizons** — I'd go with your option (b): warm up with **GT future egomotion**, then scheduled-sample toward the **planned trajectory** for train/test consistency. Note @m-zain-khawaja confirmed in #13 we're standardizing on **1 Hz prediction with horizons {1, 2, 3, 4} s**, so we can align the warp steps exactly to those buckets.
4. **Data** — the warp needs *future* ego motion over `[t, t+h]`, which is the same future-window export still open in #13 (`data_processing` currently stores only `trajectory_target`). I'll coordinate with @m-zain-khawaja to export **both the future frames** (JEPA targets) **and the future egomotion vector** (the warp) so this can be wired directly into `train_il` (#78).
5. **Target alignment** (your Q1): yes, the frozen/EMA target should be expressed in the **future ego frame** too, so prediction and target share a frame — otherwise the warp helps the input but not the supervision.
6. **History tier & memory** — start Tier-1 (done), defer Tier-2 (ConvGRU per-cell motion). On the forecast-memory knob from #13/#80: the warp **composes with any `FeatureCompressor` mode** — `occupancy` (1-channel, ≈ the channel-mean) makes the warp trivially cheap, but since the warp's value is anchoring a *rich* representation geometrically, `projected` (a learned low-rank map at the same budget) + warp is the sweet spot (cheap + semantic + grounded). I'd keep the default tied to whatever we settle in the #13 thread rather than hard-committing to 1-channel here.

Loss weighting: agree, keep equal horizon weights for now (per @m-zain-khawaja in #13). Want to take it from here, or should we pair on it?

> **Sequencing:** since the slow model is now multi-view (#85), this ego-warp lands once we confirm where the BEV future-state sits in `Reactive_E2E` post-refactor. I'll rebase the prototype branch (`feat/ego-warp-prototype`) onto `f484156` when that's settled — no rush to merge ahead of that.
